packages/x-date-pickers/src/internals/utils/date-time-utils.ts

```diff
@@ -30,7 +35,7 @@ export const resolveDateTimeFormat = <TDate extends PickerValidDate>(
   views.forEach((view) => {
     if (isTimeView(view)) {
       timeViews.push(view as TimeView);
-    } else {
+    } else if (isDatePickerView(view)) {
```

Comment:
The problem was an oversight that we did not exclude `meridiem` view, which caused `resolveDateFormat` to always return `keyboardDate` in case of a locale with 12h clock.
On 24h clock the behavior was "correct".
I've adjusted the code to avoid this inconsistency and also added an exception to not resolve `dateFormat` on `DateTimeRange` pickers as they do not have different `date` views other than `day` for now.